find/finder.go

```diff
@@ -113,9 +113,68 @@ func (f *Finder) findRoot(ctx context.Context, root *list.Element, parts []strin
 	return true
 }
 
+var refTypeMap = map[string]string{
+	"datacenter":  "Datacenter",
+	"datastore":   "Datastore",
+	"domain":      "ClusterComputeResource",
+	"dvportgroup": "DistributedVirtualPortgroup",
+	"dvs":         "DistributedVirtualSwitch",
+	"group":       "Folder",
+	"host":        "HostSystem",
+	"network":     "Network",
+	"resgroup":    "ResourcePool",
+	"vm":          "VirtualMachine",
+}
+
+var prefixTypeMap = map[string]struct{ prefix, kind string }{
+	"domain":   {"s", "ComputeResource"},
+	"group":    {"p", "StoragePod"},
+	"resgroup": {"v", "VirtualApp"},
+}
+
+func referenceFromString(arg string) *types.ManagedObjectReference {
+	if ref := object.ReferenceFromString(arg); ref != nil {
+		return ref
+	}
+
+	id := strings.SplitN(arg, "-", 2)
+	if len(id) != 2 {
+		return nil
+	}
+
+	if kind, ok := refTypeMap[id[0]]; ok {
+		if s, ok := prefixTypeMap[id[0]]; ok {
+			if strings.HasPrefix(id[1], s.prefix) {
+				return &types.ManagedObjectReference{
+					Type:  s.kind,
+					Value: arg,
+				}
+			}
+		}
+
+		return &types.ManagedObjectReference{
+			Type:  kind,
+			Value: arg,
+		}
+	}
+
+	return nil
+}
+
 func (f *Finder) find(ctx context.Context, arg string, s *spec) ([]list.Element, error) {
```

Comment:
a question here is: do we want to make it a standard behavior on finder, or should we extend its configuration to at least make the consumer explicitly state that searching by MOID is also desired?

An example: I was testing on vsphere that one can call the datacenter as "Datacenter:datacenter-3" and this is a valid name (tho I don't know which kind of person would call a datacenter on a moid format...). 

In this case, I may say I want to search my datacenter named "Datacenter:datacenter-3" but because we assume this is a moid it will fallback into a bad path.

Reply:
The call to `InventoryPath` will fail with `ManagedObjectNotFound` fault in that case, and we fall through to the name based lookup.

Reply:
I just saw `datacenter-3` is also a valid datacenter-3, `Datacenter:datacenter-3` is less likely to be used since I don't see anyone use `:` inside their name so far. But if x-x is more likely to be a often used valid name? it could be a case that one datacenter has name `datacenter-3`, and another datacenter `datacenter-4` has moid `datacenter-3`, Then the client that use this func might find the datacenter that it wants to use is unexpected.

Reply:
@lubronzhan in that case the Finder call to `InventoryPath` results in `ManagedObjectNotFound` and we fall through to name based lookup. Clients shouldn't depend on the `object.ReferenceFromString` result alone, it needs to be validated as Finder does.

Reply:
Here's an example:
```console
% govc datacenter.create datacenter-123

% govc find -l -i / -type d
Datacenter:datacenter-2   /DC0
Datacenter:datacenter-74  /datacenter-123

% govc datacenter.info datacenter-123
Name:                datacenter-123
  Path:              /datacenter-123
  Hosts:             0
  ...
  
% govc datacenter.info Datacenter:datacenter-123
govc: datacenter 'Datacenter:datacenter-123' not found

% govc datacenter.info datacenter-74
Name:                datacenter-123
  Path:              /datacenter-123
  Hosts:             0
  ...
```